Plugin not working properly with new Android Studio version

Hello!

Thanks again for the effort you put towards making this plugin!

Unfortunately, I have a problem and the plugin is not working with the new `Android Studio Meerkat | 2024.3.1 Canary`  version. Previously we had the option to start the matrix selector from `Build > Build Variant Matrix Selector` but the option is no longer there. 
Also, if I manually start the plugin by using `Run Everything` function from Android Studio and try to change the build variant using the matrix selector I am getting many errors that my modules can't not resolved.

Hi and thanks for the quick reaction!

Here's what I got from the log:

```
2024-11-29 00:38:09,819 [559864678] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - Unable to select build variant:
Cannot find module 'MyApp.feature.dashboard.impl'.
java.lang.Throwable: Unable to select build variant:
Cannot find module 'MyApp.feature.dashboard.impl'.
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
	at com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater.logAndShowBuildVariantFailure(LegacyBuildVariantUpdater.java:305)
	at com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater.updateSelectedBuildVariant(LegacyBuildVariantUpdater.java:65)
	at com.nilsenlabs.flavormatrix.actions.SelectBuildVariantMatrixAction.actionPerformed(SelectBuildVariantMatrixAction.kt:30)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.doPerformActionOrShowPopup(ActionUtil.kt:374)
	at com.intellij.ide.actions.GotoActionAction.lambda$performActionImpl$4(GotoActionAction.java:89)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.performWithActionCallbacks(ActionManagerImpl.kt:1177)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.kt:396)
	at com.intellij.ide.actions.GotoActionAction.performActionImpl(GotoActionAction.java:88)
	at com.intellij.ide.actions.GotoActionAction.lambda$performAction$2(GotoActionAction.java:68)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:236)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:25)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:198)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runIntendedWriteActionOnCurrentThread$lambda$2(AnyThreadWriteThreadingSupport.kt:217)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteIntentReadAction(AnyThreadWriteThreadingSupport.kt:128)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runIntendedWriteActionOnCurrentThread(AnyThreadWriteThreadingSupport.kt:216)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:842)
	at com.intellij.openapi.application.impl.ApplicationImpl$2.run(ApplicationImpl.java:421)
	at com.intellij.util.concurrency.ChildContext$runInChildContext$1.invoke(propagation.kt:101)
	at com.intellij.util.concurrency.ChildContext$runInChildContext$1.invoke(propagation.kt:101)
	at com.intellij.util.concurrency.ChildContext.runInChildContext(propagation.kt:107)
	at com.intellij.util.concurrency.ChildContext.runInChildContext(propagation.kt:101)
	at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:27)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:117)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:43)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
	at java.desktop/java.awt.EventQueue$4.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.desktop/java.awt.EventQueue.dispatchEvent(Unknown Source)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:675)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:573)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$18$lambda$17$lambda$16$lambda$15(IdeEventQueue.kt:355)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:857)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$18$lambda$17$lambda$16(IdeEventQueue.kt:354)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$2$lambda$1(IdeEventQueue.kt:1045)
	at com.intellij.openapi.application.WriteIntentReadAction.lambda$run$0(WriteIntentReadAction.java:24)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteIntentReadAction(AnyThreadWriteThreadingSupport.kt:128)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteIntentReadAction(ApplicationImpl.java:916)
	at com.intellij.openapi.application.WriteIntentReadAction.compute(WriteIntentReadAction.java:55)
	at com.intellij.openapi.application.WriteIntentReadAction.run(WriteIntentReadAction.java:23)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$2(IdeEventQueue.kt:1045)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$3(IdeEventQueue.kt:1054)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:109)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:1054)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$18(IdeEventQueue.kt:349)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:395)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.desktop/java.awt.EventDispatchThread.run(Unknown Source)
2024-11-29 00:38:09,820 [559864679] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - Android Studio Meerkat | 2024.3.1 Canary 3  Build #AI-243.21565.193.2431.12691553
2024-11-29 00:38:09,820 [559864679] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - JDK: 21.0.5; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o.
2024-11-29 00:38:09,820 [559864679] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - OS: Mac OS X
2024-11-29 00:38:09,820 [559864679] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - Plugin to blame: Build Variant Matrix Selector version: 1.3.0
2024-11-29 00:38:09,820 [559864679] SEVERE - #com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater - Last Action: SelectBuildVariantMatrixAction
```

I am getting this error for all the modules. When I click OK on all the error dialogs the plugin actually tries to change the build variant but changes it successfully only for some and not all of them.

--- a/DEVNOTES.txt
+++ b/DEVNOTES.txt
@@ -1,0 +1,10 @@
+## Issues with Java compile version
+
+when attempting `jar` execution and you get "Unsupported class file major version 65":
+
+- Build & Execution
+  - Build Tools
+    - Gradle
+      - Gradle Jvm -> temurin 17
+
+Might have to switch to 21 when running locally against IDE

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ plugins {
 }
 
 group 'com.nilsenlabs.flavormatrix'
-version '1.4.0-SNAPSHOT'
+version '1.4.0-alpha1'
 
 Properties properties = new Properties()
 properties.load(project.rootProject.file("local.properties").newDataInputStream())

--- a/build.gradle
+++ b/build.gradle
@@ -7,7 +7,7 @@ buildscript {
 }
 
 plugins {
-    id 'org.jetbrains.kotlin.jvm' version '1.8.0'
+    id 'org.jetbrains.kotlin.jvm' version '2.0.0'
     id 'org.jetbrains.intellij' version '1.13.3'
 }
 

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ plugins {
 }
 
 group 'com.nilsenlabs.flavormatrix'
-version '1.4.0-alpha2'
+version '1.4.0'
 
 Properties properties = new Properties()
 properties.load(project.rootProject.file("local.properties").newDataInputStream())

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ plugins {
 }
 
 group 'com.nilsenlabs.flavormatrix'
-version '1.4.0-alpha1'
+version '1.4.0-alpha2'
 
 Properties properties = new Properties()
 properties.load(project.rootProject.file("local.properties").newDataInputStream())

--- a/build.gradle
+++ b/build.gradle
@@ -31,7 +31,7 @@ def defaultLocalStudioPath = 'd:/coding/android-studio-meerkat'
 // Also see https://plugins.jetbrains.com/docs/intellij/tools-gradle-intellij-plugin.html#intellij-extension-localpath
 def localStudioPath = properties.get("localStudioPath") ?: defaultLocalStudioPath
 
-// Below: Meerkat 2024.3.1 Canary 3
+// Below: Meerkat 2024.3.1 Canary 3. Remember to update plugin.xml
 def targetIdeVersion = "243.21565.193"
 
 intellij {

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ plugins {
 }
 
 group 'com.nilsenlabs.flavormatrix'
-version '1.3.0'
+version '1.4.0-SNAPSHOT'
 
 Properties properties = new Properties()
 properties.load(project.rootProject.file("local.properties").newDataInputStream())
@@ -24,15 +24,15 @@ properties.load(project.rootProject.file("local.properties").newDataInputStream(
 //  plugins\properties\lib\platform-images.jar
 //  plugins\properties\lib\properties.jar
 //  plugins\smali\lib\smali.jar
-def defaultLocalStudioPath = 'd:/coding/android-studio-ladybug-canary1'
+def defaultLocalStudioPath = 'd:/coding/android-studio-meerkat'
 
 // Specify the path in your local.properties file as
 // localStudioPath=your/path/to/androidstudio
 // Also see https://plugins.jetbrains.com/docs/intellij/tools-gradle-intellij-plugin.html#intellij-extension-localpath
 def localStudioPath = properties.get("localStudioPath") ?: defaultLocalStudioPath
 
-// Below: Ladybug 2024.1.3 Canary 1
-def targetIdeVersion = "241.18034.62"
+// Below: Meerkat 2024.3.1 Canary 3
+def targetIdeVersion = "243.21565.193"
 
 intellij {
     // Uncomment the two lines below and remove "localPath" if you want to build for a different version than the one locally installed
@@ -50,6 +50,7 @@ intellij {
 patchPluginXml {
     changeNotes = """
     <ul>
+      <li>1.4.0 Support for Android Studio Meerkat (from Canary 3)</li>
       <li>1.3.0 Support for Android Studio Ladybug</li>
       <li>1.2.0 Support for Android Studio Giraffe and Hedgehog</li>
       <li>1.1.0 Support for Android Studio Electric Eel</li>

--- a/src/main/java/com/nilsenlabs/flavormatrix/actions/AndroidModuleHelper.kt
+++ b/src/main/java/com/nilsenlabs/flavormatrix/actions/AndroidModuleHelper.kt
@@ -1,11 +1,12 @@
 package com.nilsenlabs.flavormatrix.actions
 
 import com.android.tools.idea.gradle.project.model.GradleAndroidModel
+import com.android.tools.idea.projectsystem.gradle.getHolderModule
 import com.intellij.openapi.module.Module
 import kotlin.streams.toList
 
 object AndroidModuleHelper {
-    fun createDimensionTable(androidModules: List<GradleAndroidModel>, modules: Array<Module> ): DimensionList {
+    fun createDimensionTable(androidModules: List<GradleAndroidModel>, modules: Array<Module>): DimensionList {
         val dimensions = createMergedDimensionList(androidModules)
         addBuildTypes(modules, dimensions)
         dimensions.createOrderedDimensionMaps(androidModules)
@@ -32,7 +33,7 @@ object AndroidModuleHelper {
         for (module in modules) {
             val flavors = module.androidProject.multiVariantData?.productFlavors?.toList() ?: continue
             for (flavorObj in flavors) {
-                // println("Processing flavorName: ${flavorObj.productFlavor.name}: ${flavorObj.productFlavor.dimension}")
+                // getLog.info("Processing flavorName: ${flavorObj.productFlavor.name}: ${flavorObj.productFlavor.dimension}")
                 /*
                     Expecting a log like like this for my test project:
                     Processing flavorName: qa: environment
@@ -53,14 +54,12 @@ object AndroidModuleHelper {
 
     fun getBuildTypes(modules: Array<Module>): List<String> {
         val buildTypes = mutableSetOf<String>()
-        for(module in modules) {
-
+        for (module in modules) {
             val amod = GradleAndroidModel.get(module)
-            System.out.println( "Selected variant for ${module.name}: ${amod?.selectedVariant?.displayName} -- ${amod?.selectedVariant?.buildType} -- ${amod?.selectedVariant?.productFlavors?.joinToString("*")}")
             buildTypes.addAll(amod?.buildTypeNames ?: emptyList())
         }
         val list = buildTypes.stream().toList()
-        System.out.println("List returned: $list ")
+        getLog().info("Build type list resolved after iterating ${modules.size} modules: $list")
         return list
     }
 }

--- a/src/main/java/com/nilsenlabs/flavormatrix/actions/Common.kt
+++ b/src/main/java/com/nilsenlabs/flavormatrix/actions/Common.kt
@@ -1,0 +1,7 @@
+package com.nilsenlabs.flavormatrix.actions
+
+import com.intellij.openapi.diagnostic.Logger
+
+fun getLog(): Logger {
+    return Logger.getInstance(SelectBuildVariantMatrixAction::class.java)
+}

--- a/src/main/java/com/nilsenlabs/flavormatrix/actions/SelectBuildVariantMatrixAction.kt
+++ b/src/main/java/com/nilsenlabs/flavormatrix/actions/SelectBuildVariantMatrixAction.kt
@@ -3,6 +3,8 @@ package com.nilsenlabs.flavormatrix.actions
 import com.android.tools.idea.gradle.project.model.GradleAndroidModel
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.module.Module
 import com.intellij.openapi.module.ModuleManager
 import com.nilsenlabs.flavormatrix.actions.old.LegacyBuildVariantUpdater
 
@@ -13,21 +15,39 @@ class SelectBuildVariantMatrixAction : AnAction() {
         val legacyUpdater = LegacyBuildVariantUpdater()
         val moduleManager = ModuleManager.getInstance(project)
 
-        val androidModules = moduleManager.modules
-            .map { GradleAndroidModel.get(it) }
-            .filter { it?.moduleName != null }
-            .map { it!! }
+        val unfilteredModuleList = moduleManager.modules
+            .map { module ->
+                GradleAndroidModel.get(module)?.let { Pair(it, module) }
+            }
+            .filterNotNull()
+
+        getLog().info("Module list before filtering:")
+        for (pair in unfilteredModuleList) {
+            getLog().info("AndroidModule: ${pair.first.moduleName} -> PlainModule: ${pair.second.name}")
+        }
+
+        val androidModulePairs: Map<GradleAndroidModel?, Module> = unfilteredModuleList
+            .filter { it.first.moduleName.lowercase() == it.second.name.lowercase() }
             .distinct()
+            .toMap()
+        val androidModules = androidModulePairs.keys.filterNotNull()
+
+        getLog().info("Module list after filtering:")
+        for (pair in androidModulePairs) {
+            getLog().info("AndroidModule: ${pair.key?.moduleName} -> PlainModule: ${pair.value.name}")
+        }
 
         val dimensions = AndroidModuleHelper.createDimensionTable(androidModules, moduleManager.modules)
 
+        getLog().info("Staring variant update...")
         val dialog = VariantSelectorDialog(dimensions, androidModules, project)
         if (dialog.showAndGet()) {
             // OK selected => Post variant selection back to Android Studio
-            for (andModule in androidModules) {
-                dimensions.getSelectedVariantFor(andModule.moduleName)?.let {vari ->
-                    println("Found module ${andModule.moduleName}. Updating to variant $vari")
-                    legacyUpdater.updateSelectedBuildVariant(project, andModule.moduleName, vari)
+            for (modulePair in androidModulePairs) {
+                val andModule = modulePair.key ?: continue // it being null should not happen
+                dimensions.getSelectedVariantFor(andModule.moduleName)?.let { vari ->
+                    getLog().info("Module: ${andModule.moduleName}. Updating to variant $vari")
+                    legacyUpdater.updateSelectedBuildVariant(project, modulePair.value, vari)
                 }
             }
             LegacyBuildVariantUpdater.requestGradleSync(project)

--- a/src/main/java/com/nilsenlabs/flavormatrix/actions/old/LegacyBuildVariantUpdater.java
+++ b/src/main/java/com/nilsenlabs/flavormatrix/actions/old/LegacyBuildVariantUpdater.java
@@ -59,25 +59,22 @@ public class LegacyBuildVariantUpdater {
         this.mySelectionChangeListeners.add(listener);
     }
 
-    public boolean updateSelectedBuildVariant(@NotNull Project project, @NotNull String moduleName, @NotNull String selectedBuildVariant) {
-        Module moduleToUpdate = findModule(project, moduleName);
-        if (moduleToUpdate == null) {
-            logAndShowBuildVariantFailure(String.format("Cannot find module '%1$s'.", moduleName));
-            return false;
-        } else {
-            NdkModuleModel ndkModuleModel = getNdkModelIfItHasNativeVariantAbis(moduleToUpdate);
-            NdkFacet ndkFacet = NdkFacet.getInstance(moduleToUpdate);
-            if (ndkModuleModel != null && ndkFacet != null) {
-                VariantAbi newVariantAbi = resolveNewVariantAbi(ndkFacet, ndkModuleModel, selectedBuildVariant, (String)null);
-                if (newVariantAbi == null) {
-                    logAndShowBuildVariantFailure(String.format("Cannot find suitable ABI for native module '%1$s'.", moduleName));
-                    return false;
-                } else {
-                    return this.updateSelectedVariant(project, moduleName, VariantAndAbi.fromVariantAbi(newVariantAbi));
-                }
+    public boolean updateSelectedBuildVariant(@NotNull Project project, Module moduleToUpdate, @NotNull String selectedBuildVariant) {
+        String moduleName = moduleToUpdate.getName();
+        NdkModuleModel ndkModuleModel = getNdkModelIfItHasNativeVariantAbis(moduleToUpdate);
+        NdkFacet ndkFacet = NdkFacet.getInstance(moduleToUpdate);
+        if (ndkModuleModel != null && ndkFacet != null) {
+            // NDK
+            VariantAbi newVariantAbi = resolveNewVariantAbi(ndkFacet, ndkModuleModel, selectedBuildVariant, (String)null);
+            if (newVariantAbi == null) {
+                logAndShowBuildVariantFailure(String.format("Cannot find suitable ABI for native module '%1$s'.", moduleName));
+                return false;
             } else {
-                return this.updateSelectedVariant(project, moduleName, new VariantAndAbi(selectedBuildVariant, (String)null));
+                return this.updateSelectedVariant(project, moduleToUpdate, VariantAndAbi.fromVariantAbi(newVariantAbi));
             }
+        } else {
+            // Android
+            return this.updateSelectedVariant(project, moduleToUpdate, new VariantAndAbi(selectedBuildVariant, (String)null));
         }
     }
 
@@ -102,7 +99,7 @@ public class LegacyBuildVariantUpdater {
                     logAndShowAbiNameFailure(String.format("Cannot find suitable ABI for native module '%1$s'.", moduleName));
                     return false;
                 } else {
-                    return this.updateSelectedVariant(project, moduleName, VariantAndAbi.fromVariantAbi(newVariantAbi));
+                    return this.updateSelectedVariant(project, moduleToUpdate, VariantAndAbi.fromVariantAbi(newVariantAbi));
                 }
             } else {
                 logAndShowAbiNameFailure(String.format("Cannot find native module model '%1$s'.", moduleName));
@@ -111,12 +108,8 @@ public class LegacyBuildVariantUpdater {
         }
     }
 
-    private boolean updateSelectedVariant(@NotNull Project project, @NotNull String moduleName, @NotNull VariantAndAbi variantAndAbi) {
-        Module module = findModule(project, moduleName);
-        if (module == null) {
-            logAndShowBuildVariantFailure(String.format("Cannot find module '%1$s'.", moduleName));
-            return false;
-        } else if (!findAndUpdateAffectedFacets(module, variantAndAbi)) {
+    private boolean updateSelectedVariant(@NotNull Project project, @NotNull Module module, @NotNull VariantAndAbi variantAndAbi) {
+        if (!findAndUpdateAffectedFacets(module, variantAndAbi)) {
             return false;
         } else {
             ExternalProjectInfo data = ProjectDataManager.getInstance().getExternalProjectData(project, GradleConstants.SYSTEM_ID, project.getBasePath());

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -17,7 +17,7 @@
     <id>com.nilsenlabs.flavormatrix</id>
     <name>Build Variant Matrix Selector</name>
     <vendor url="https://github.com/Nilzor/build-variant-matrix">Frode Nilsen</vendor>
-    <idea-version since-build="241.18034.62" />
+    <idea-version since-build="243.21565.193" />
     <description><![CDATA[<p>Select variant by selecting flavors in a matrix instead of through the drop down lists provdided by IntelliJ/Android Studio
 in the "Build Variants" view. In addition this plugin selects variants for all modules at the same time, even
 when there are multiple leaf modules. This results in a very fast and intuitive way of switching variants

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -44,7 +44,7 @@ This is an open source project licensed under the Apache License v2.0.]]></descr
         <action id="SelectBuildVariantMatrixAction"
                 class="com.nilsenlabs.flavormatrix.actions.SelectBuildVariantMatrixAction"
                 text="Select Build Variant by Matrix">
-            <add-to-group group-id="GradleProjectStructureActions"
+            <add-to-group group-id="BuildMenu"
                           anchor="after"
                           relative-to-action="SelectBuildVariant" />
             <keyboard-shortcut keymap="$default" first-keystroke="ctrl shift x"/>

